// appium-ios-simulator/build/lib/calendar.js
'use strict';

var _createClass = require('babel-runtime/helpers/create-class')['default'];

var _classCallCheck = require('babel-runtime/helpers/class-call-check')['default'];

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _appiumSupport = require('appium-support');

var _utils = require('./utils');

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _logger = require('./logger');

var _logger2 = _interopRequireDefault(_logger);

var Calendar = (function () {
  function Calendar(xcodeVersion, sharedResourcesDir) {
    _classCallCheck(this, Calendar);

    this.xcodeVersion = xcodeVersion;
    this.sharedResourcesDir = sharedResourcesDir;
  }

  _createClass(Calendar, [{
    key: 'getDB',
    value: function getDB() {
      var tccPath;
      return _regeneratorRuntime.async(function getDB$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            if (!this.db) {
              context$2$0.next = 2;
              break;
            }

            return context$2$0.abrupt('return', this.db);

          case 2:
            tccPath = _path2['default'].resolve(this.sharedResourcesDir, 'Library', 'TCC');
            context$2$0.next = 5;
            return _regeneratorRuntime.awrap(_appiumSupport.fs.exists(tccPath));

          case 5:
            if (context$2$0.sent) {
              context$2$0.next = 8;
              break;
            }

            context$2$0.next = 8;
            return _regeneratorRuntime.awrap((0, _appiumSupport.mkdirp)(tccPath));

          case 8:

            this.db = _path2['default'].resolve(tccPath, 'TCC.db');
            context$2$0.next = 11;
            return _regeneratorRuntime.awrap((0, _utils.execSQLiteQuery)(this.db, 'CREATE TABLE IF NOT EXISTS access (\n      service TEXT NOT NULL DEFAULT \'\',\n      client TEXT NOT NULL DEFAULT \'\',\n      client_type INTEGER,\n      allowed INTEGER,\n      prompt_count INTEGER,\n      csreq BLOB NOT NULL DEFAULT \'\',\n      policy_ID INTEGER,\n      PRIMARY KEY(service, client, client_type)\n    );'.replace(/\n/g, '')));

          case 11:
            return context$2$0.abrupt('return', this.db);

          case 12:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'getCalendarRowCount',
    value: function getCalendarRowCount(bundleID) {
      var db, count;
      return _regeneratorRuntime.async(function getCalendarRowCount$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            context$2$0.next = 2;
            return _regeneratorRuntime.awrap(this.getDB());

          case 2:
            db = context$2$0.sent;
            context$2$0.next = 5;
            return _regeneratorRuntime.awrap((0, _utils.execSQLiteQuery)(db, 'SELECT count(*) FROM access WHERE client=\'?\' AND service=\'kTCCServiceCalendar\';', bundleID));

          case 5:
            count = context$2$0.sent;
            return context$2$0.abrupt('return', parseInt(count.stdout.split('=')[1], 10));

          case 7:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'hasCalendarAccess',
    value: function hasCalendarAccess(bundleID) {
      var count;
      return _regeneratorRuntime.async(function hasCalendarAccess$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            context$2$0.t0 = _regeneratorRuntime;
            context$2$0.next = 3;
            return _regeneratorRuntime.awrap(this.getDB());

          case 3:
            context$2$0.t1 = context$2$0.sent;
            context$2$0.t2 = bundleID;
            context$2$0.t3 = (0, _utils.execSQLiteQuery)(context$2$0.t1, 'SELECT count(*) FROM access WHERE client=\'?\' AND allowed=1 AND service=\'kTCCServiceCalendar\';', context$2$0.t2);
            context$2$0.next = 8;
            return context$2$0.t0.awrap.call(context$2$0.t0, context$2$0.t3);

          case 8:
            count = context$2$0.sent;
            return context$2$0.abrupt('return', parseInt(count.stdout.split('=')[1], 10) > 0);

          case 10:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'enableCalendarAccess',
    value: function enableCalendarAccess(bundleID) {
      var db, query, count, _ref, stdout;

      return _regeneratorRuntime.async(function enableCalendarAccess$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            context$2$0.next = 2;
            return _regeneratorRuntime.awrap(this.getDB());

          case 2:
            db = context$2$0.sent;
            query = undefined;
            context$2$0.next = 6;
            return _regeneratorRuntime.awrap(this.getCalendarRowCount(bundleID));

          case 6:
            context$2$0.t0 = context$2$0.sent;

            if (!(context$2$0.t0 > 0)) {
              context$2$0.next = 11;
              break;
            }

            query = 'UPDATE \'access\' SET\n        service=\'kTCCServiceCalendar\',\n        client_type=0,\n        allowed=1,\n        prompt_count=1,\n        csreq=0\n      WHERE client=\'?\' AND service=\'kTCCServiceCalendar\''.replace(/\n/g, ' ');
            context$2$0.next = 26;
            break;

          case 11:
            count = this.columnCount;

            if (_lodash2['default'].isNumber(this.columnCount)) {
              context$2$0.next = 25;
              break;
            }

            context$2$0.prev = 13;
            context$2$0.next = 16;
            return _regeneratorRuntime.awrap((0, _utils.execSQLiteQuery)(db, 'pragma table_info(\'access\')'));

          case 16:
            _ref = context$2$0.sent;
            stdout = _ref.stdout;

            // stdout has a number of records, each starting with `cid`
            //       cid = 0
            //       name = service
            //       type = TEXT
            //       notnull = 1
            //       dflt_value =
            //       pk = 1
            //
            //       cid = 1
            //       name = client
            //       type = TEXT
            //       notnull = 1
            //       dflt_value =
            //       pk = 2
            count = this.columnCount = stdout.trim()
            // get each record's beginning
            .split('cid')
            // remove the empty row (usually, the first)
            .filter(function (record) {
              return !_lodash2['default'].isEmpty(record);
            }).length;
            context$2$0.next = 25;
            break;

          case 21:
            context$2$0.prev = 21;
            context$2$0.t1 = context$2$0['catch'](13);

            // use defaults, but do not save
            count = this.xcodeVersion.major < 10 ? 7 : 11;
            _logger2['default'].warn('Unable to find Calendar access column count: ' + context$2$0.t1.message);

          case 25:

            // fill in the query with 0's for columns after the 5 we care about
            query = 'INSERT INTO \'access\' VALUES (\'kTCCServiceCalendar\', \'?\', 0, 1, 1, ' + _lodash2['default'].range(0, count - 5, 0).join(', ') + ');';

          case 26:
            context$2$0.next = 28;
            return _regeneratorRuntime.awrap((0, _utils.execSQLiteQuery)(db, query, bundleID));

          case 28:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this, [[13, 21]]);
    }
  }, {
    key: 'disableCalendarAccess',
    value: function disableCalendarAccess(bundleID) {
      var db;
      return _regeneratorRuntime.async(function disableCalendarAccess$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            context$2$0.next = 2;
            return _regeneratorRuntime.awrap(this.getDB());

          case 2:
            db = context$2$0.sent;
            context$2$0.next = 5;
            return _regeneratorRuntime.awrap(this.getCalendarRowCount(bundleID));

          case 5:
            context$2$0.t0 = context$2$0.sent;

            if (!(context$2$0.t0 > 0)) {
              context$2$0.next = 9;
              break;
            }

            context$2$0.next = 9;
            return _regeneratorRuntime.awrap((0, _utils.execSQLiteQuery)(db, 'DELETE FROM \'access\' WHERE client=\'?\' AND service=\'kTCCServiceCalendar\''.replace(/\n/g, ' '), bundleID));

          case 9:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }]);

  return Calendar;
})();

exports['default'] = Calendar;
module.exports = exports['default'];
