// appium-tizen-driver/build/index.js
#!/usr/bin/env node

require('source-map-support').install();

'use strict';

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _yargs = require('yargs');

var _yargs2 = _interopRequireDefault(_yargs);

var _asyncbox = require('asyncbox');

var _libServer = require('./lib/server');

var _libDriver = require('./lib/driver');

var _libDriver2 = _interopRequireDefault(_libDriver);

var _libTizenHelpers = require('./lib/tizen-helpers');

var _libTizenHelpers2 = _interopRequireDefault(_libTizenHelpers);

var _libCommandsIndex = require('./lib/commands/index');

var _libCommandsIndex2 = _interopRequireDefault(_libCommandsIndex);

var _libDesiredCaps = require('./lib/desired-caps');

var DEFAULT_HOST = "localhost";
var DEFAULT_PORT = 4723;

function main() {
  var port, host;
  return _regeneratorRuntime.async(function main$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        port = _yargs2['default'].argv.port || DEFAULT_PORT;
        host = _yargs2['default'].argv.host || DEFAULT_HOST;
        return context$1$0.abrupt('return', (0, _libServer.startServer)(port, host));

      case 3:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

if (require.main === module) {
  (0, _asyncbox.asyncify)(main);
}

exports['default'] = _libDriver2['default'];
exports.tizenHelpers = _libTizenHelpers2['default'];
exports.tizenCommands = _libCommandsIndex2['default'];
exports.TizenDriver = _libDriver2['default'];
exports.startServer = _libServer.startServer;
exports.commonCapConstraints = _libDesiredCaps.commonCapConstraints;
