// appium-base-driver/build/index.js
require('source-map-support').install();

// BaseDriver exports
'use strict';

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
         value: true
});

var _libBasedriverDriver = require('./lib/basedriver/driver');

var _libBasedriverDriver2 = _interopRequireDefault(_libBasedriverDriver);

var _libBasedriverImageElement = require('./lib/basedriver/image-element');

var _libBasedriverDeviceSettings = require('./lib/basedriver/device-settings');

var _libBasedriverDeviceSettings2 = _interopRequireDefault(_libBasedriverDeviceSettings);

// MJSONWP exports

var _libProtocol = require('./lib/protocol');

// Express exports

var _libExpressStatic = require('./lib/express/static');

var _libExpressServer = require('./lib/express/server');

// jsonwp-proxy exports

var _libJsonwpProxyProxy = require('./lib/jsonwp-proxy/proxy');

var _libJsonwpProxyProxy2 = _interopRequireDefault(_libJsonwpProxyProxy);

// jsonwp-status exports

var _libJsonwpStatusStatus = require('./lib/jsonwp-status/status');

// W3C capabilities parser

var _libBasedriverCapabilities = require('./lib/basedriver/capabilities');

// Web socket helpers

var _libExpressWebsocket = require('./lib/express/websocket');

exports.BaseDriver = _libBasedriverDriver2['default'];
exports.DeviceSettings = _libBasedriverDeviceSettings2['default'];
exports.ImageElement = _libBasedriverImageElement.ImageElement;
exports.BASEDRIVER_HANDLED_SETTINGS = _libBasedriverDeviceSettings.BASEDRIVER_HANDLED_SETTINGS;
exports['default'] = _libBasedriverDriver2['default'];
exports.Protocol = _libProtocol.Protocol;
exports.routeConfiguringFunction = _libProtocol.routeConfiguringFunction;
exports.errors = _libProtocol.errors;
exports.isErrorType = _libProtocol.isErrorType;
exports.errorFromMJSONWPStatusCode = _libProtocol.errorFromMJSONWPStatusCode;
exports.errorFromW3CJsonCode = _libProtocol.errorFromW3CJsonCode;
exports.errorFromCode = _libProtocol.errorFromMJSONWPStatusCode;
exports.ALL_COMMANDS = _libProtocol.ALL_COMMANDS;
exports.METHOD_MAP = _libProtocol.METHOD_MAP;
exports.routeToCommandName = _libProtocol.routeToCommandName;
exports.NO_SESSION_ID_COMMANDS = _libProtocol.NO_SESSION_ID_COMMANDS;
exports.isSessionCommand = _libProtocol.isSessionCommand;
exports.STATIC_DIR = _libExpressStatic.STATIC_DIR;
exports.server = _libExpressServer.server;
exports.JWProxy = _libJsonwpProxyProxy2['default'];

var statusCodes = _libJsonwpStatusStatus.codes;
exports.statusCodes = statusCodes;
exports.getSummaryByCode = _libJsonwpStatusStatus.getSummaryByCode;
exports.processCapabilities = _libBasedriverCapabilities.processCapabilities;
exports.DEFAULT_WS_PATHNAME_PREFIX = _libExpressWebsocket.DEFAULT_WS_PATHNAME_PREFIX;
