// appium-tizen-driver/build/lib/driver.js
'use strict';

var _get = require('babel-runtime/helpers/get')['default'];

var _inherits = require('babel-runtime/helpers/inherits')['default'];

var _createClass = require('babel-runtime/helpers/create-class')['default'];

var _classCallCheck = require('babel-runtime/helpers/class-call-check')['default'];

var _slicedToArray = require('babel-runtime/helpers/sliced-to-array')['default'];

var _getIterator = require('babel-runtime/core-js/get-iterator')['default'];

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _Object$assign = require('babel-runtime/core-js/object/assign')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _appiumBaseDriver = require('appium-base-driver');

var _desiredCaps = require('./desired-caps');

var _desiredCaps2 = _interopRequireDefault(_desiredCaps);

var _commandsIndex = require('./commands/index');

var _commandsIndex2 = _interopRequireDefault(_commandsIndex);

var _tizenHelpers = require('./tizen-helpers');

var _tizenHelpers2 = _interopRequireDefault(_tizenHelpers);

var _tizenBootstrapJs = require('./tizen-bootstrap.js');

var _tizenBootstrapJs2 = _interopRequireDefault(_tizenBootstrapJs);

var _logger = require('./logger');

var _logger2 = _interopRequireDefault(_logger);

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _appiumSdb = require('appium-sdb');

var _appiumSupport = require('appium-support');

var _asyncbox = require('asyncbox');

var _packageJson = require('../../package.json');

// eslint-disable-line import/no-unresolved

var BOOTSTRAP_PORT = 8888;
var NO_PROXY = [['POST', new RegExp('^/session/[^/]+/appium')], ['GET', new RegExp('^/session/[^/]+/appium')]];

var TizenDriver = (function (_BaseDriver) {
  _inherits(TizenDriver, _BaseDriver);

  function TizenDriver() {
    var opts = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];
    var shouldValidateCaps = arguments.length <= 1 || arguments[1] === undefined ? true : arguments[1];

    _classCallCheck(this, TizenDriver);

    _get(Object.getPrototypeOf(TizenDriver.prototype), 'constructor', this).call(this, opts, shouldValidateCaps);

    _logger2['default'].debug('TizenDriver version: ' + _packageJson.version);

    this.locatorStrategies = ['id', 'accessibility id', 'class name'];

    this.desiredCapConstraints = _desiredCaps2['default'];
    this.jwpProxyActive = false;
    this.jwpProxyAvoid = _lodash2['default'].clone(NO_PROXY);
    this.settings = new _appiumBaseDriver.DeviceSettings({ ignoreUnimportantViews: false });
    this.bootstrapPort = BOOTSTRAP_PORT;

    var _iteratorNormalCompletion = true;
    var _didIteratorError = false;
    var _iteratorError = undefined;

    try {
      for (var _iterator = _getIterator(_lodash2['default'].pairs(_commandsIndex2['default'])), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
        var _step$value = _slicedToArray(_step.value, 2);

        var cmd = _step$value[0];
        var fn = _step$value[1];

        TizenDriver.prototype[cmd] = fn;
      }
    } catch (err) {
      _didIteratorError = true;
      _iteratorError = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion && _iterator['return']) {
          _iterator['return']();
        }
      } finally {
        if (_didIteratorError) {
          throw _iteratorError;
        }
      }
    }
  }

  _createClass(TizenDriver, [{
    key: 'createSession',
    value: function createSession(caps) {
      var sessionId, _ref, _ref2, serverDetails, defaultOpts, _ref3, udid, emPort;

      return _regeneratorRuntime.async(function createSession$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            context$2$0.prev = 0;
            sessionId = undefined;
            context$2$0.next = 4;
            return _regeneratorRuntime.awrap(_get(Object.getPrototypeOf(TizenDriver.prototype), 'createSession', this).call(this, caps));

          case 4:
            _ref = context$2$0.sent;
            _ref2 = _slicedToArray(_ref, 1);
            sessionId = _ref2[0];
            serverDetails = { platform: 'LINUX',
              webStorageEnabled: false,
              takesScreenshot: false,
              javascriptEnabled: true,
              databaseEnabled: false,
              networkConnectionEnabled: false,
              locationContextEnabled: false,
              warnings: {},
              desired: this.caps };

            this.caps = _Object$assign(serverDetails, this.caps);

            context$2$0.next = 11;
            return _regeneratorRuntime.awrap(_appiumSupport.tempDir.staticDir());

          case 11:
            context$2$0.t0 = context$2$0.sent;
            context$2$0.t1 = _appiumSdb.DEFAULT_SDB_PORT;
            defaultOpts = {
              tmpDir: context$2$0.t0,
              fullReset: false,
              sdbPort: context$2$0.t1,
              tizenInstallTimeout: 50000
            };

            _lodash2['default'].defaults(this.opts, defaultOpts);

            if (this.opts.noReset === true) {
              this.opts.fullReset = false;
            }
            if (this.opts.fullReset === true) {
              this.opts.noReset = false;
            }
            this.opts.fastReset = !this.opts.fullReset && !this.opts.noReset;
            this.opts.skipUninstall = this.opts.fastReset || this.opts.noReset;

            context$2$0.next = 21;
            return _regeneratorRuntime.awrap(_tizenHelpers2['default'].getDeviceInfoFromCaps(this.opts));

          case 21:
            _ref3 = context$2$0.sent;
            udid = _ref3.udid;
            emPort = _ref3.emPort;

            this.opts.udid = udid;
            this.opts.emPort = emPort;

            context$2$0.next = 28;
            return _regeneratorRuntime.awrap(_tizenHelpers2['default'].createSDB(this.opts.udid, this.opts.emPort, this.opts.sdbPort, this.opts.suppressKillServer));

          case 28:
            this.sdb = context$2$0.sent;
            context$2$0.next = 31;
            return _regeneratorRuntime.awrap(this.startTizenSession(this.opts));

          case 31:
            return context$2$0.abrupt('return', [sessionId, this.caps]);

          case 34:
            context$2$0.prev = 34;
            context$2$0.t2 = context$2$0['catch'](0);
            context$2$0.prev = 36;
            context$2$0.next = 39;
            return _regeneratorRuntime.awrap(this.deleteSession());

          case 39:
            context$2$0.next = 43;
            break;

          case 41:
            context$2$0.prev = 41;
            context$2$0.t3 = context$2$0['catch'](36);

          case 43:
            throw context$2$0.t2;

          case 44:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this, [[0, 34], [36, 41]]);
    }
  }, {
    key: 'startTizenSession',
    value: function startTizenSession() {
      var isAppInstalled, isStartedApp;
      return _regeneratorRuntime.async(function startTizenSession$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            if (!this.opts.reboot) {
              context$2$0.next = 9;
              break;
            }

            context$2$0.next = 3;
            return _regeneratorRuntime.awrap(this.sdb.root());

          case 3:
            context$2$0.next = 5;
            return _regeneratorRuntime.awrap(this.sdb.reboot());

          case 5:
            context$2$0.next = 7;
            return _regeneratorRuntime.awrap((0, _asyncbox.sleep)(40000));

          case 7:
            context$2$0.next = 9;
            return _regeneratorRuntime.awrap(this.sdb.stopAutoSleep());

          case 9:
            if (!this.opts.app) {
              context$2$0.next = 12;
              break;
            }

            context$2$0.next = 12;
            return _regeneratorRuntime.awrap(this.installApp(this.opts.app));

          case 12:
            context$2$0.next = 14;
            return _regeneratorRuntime.awrap(this.isAppInstalled(this.opts.appPackage));

          case 14:
            isAppInstalled = context$2$0.sent;

            if (!isAppInstalled) {
              _logger2['default'].errorAndThrow('Could not find to App in device.');
            }

            if (!this.opts.appPackage) {
              context$2$0.next = 25;
              break;
            }

            context$2$0.next = 19;
            return _regeneratorRuntime.awrap(this.isStartedApp());

          case 19:
            isStartedApp = context$2$0.sent;

            if (!isStartedApp) {
              context$2$0.next = 23;
              break;
            }

            context$2$0.next = 23;
            return _regeneratorRuntime.awrap(this.closeApp());

          case 23:
            context$2$0.next = 25;
            return _regeneratorRuntime.awrap(this.startApp({ timeout: 20000 }));

          case 25:

            this.bootstrap = new _tizenBootstrapJs2['default'](this.sdb, this.bootstrapPort, this.opts.websocket);
            context$2$0.next = 28;
            return _regeneratorRuntime.awrap(this.bootstrap.start(this.opts.appPackage));

          case 28:
            context$2$0.next = 30;
            return _regeneratorRuntime.awrap(this.bootstrap.pushUiAutomator());

          case 30:
            if (!this.opts.ignoreUnimportantViews) {
              context$2$0.next = 33;
              break;
            }

            context$2$0.next = 33;
            return _regeneratorRuntime.awrap(this.settings.update({ ignoreUnimportantViews: this.opts.ignoreUnimportantViews }));

          case 33:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'checkPackagePresent',
    value: function checkPackagePresent() {
      return _regeneratorRuntime.async(function checkPackagePresent$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            _logger2['default'].debug("Checking whether package is present on the device");
            context$2$0.next = 3;
            return _regeneratorRuntime.awrap(this.sdb.shell(['app_launcher --list | grep ' + this.opts.appPackage]));

          case 3:
            if (context$2$0.sent) {
              context$2$0.next = 5;
              break;
            }

            _logger2['default'].errorAndThrow('Could not find package ' + this.opts.appPackage + ' on the device');

          case 5:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'deleteSession',
    value: function deleteSession() {
      return _regeneratorRuntime.async(function deleteSession$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            _logger2['default'].debug("Shutting down Tizen driver");
            context$2$0.next = 3;
            return _regeneratorRuntime.awrap(_get(Object.getPrototypeOf(TizenDriver.prototype), 'deleteSession', this).call(this));

          case 3:
            if (!this.bootstrap) {
              context$2$0.next = 18;
              break;
            }

            context$2$0.next = 6;
            return _regeneratorRuntime.awrap(this.sdb.forceStop(this.opts.appPackage));

          case 6:
            if (!(this.opts.fullReset && !this.opts.skipUninstall && !this.appOnDevice)) {
              context$2$0.next = 9;
              break;
            }

            context$2$0.next = 9;
            return _regeneratorRuntime.awrap(this.sdb.uninstall(this.opts.appPackage));

          case 9:
            context$2$0.next = 11;
            return _regeneratorRuntime.awrap(this.sdb.startAutoSleep());

          case 11:
            context$2$0.next = 13;
            return _regeneratorRuntime.awrap(this.bootstrap.removeUiAutomator());

          case 13:
            context$2$0.next = 15;
            return _regeneratorRuntime.awrap(this.bootstrap.shutdown());

          case 15:
            this.bootstrap = null;
            context$2$0.next = 19;
            break;

          case 18:
            _logger2['default'].debug("Called deleteSession but bootstrap wasn't active");

          case 19:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'validateDesiredCaps',
    value: function validateDesiredCaps(caps) {
      var res = _get(Object.getPrototypeOf(TizenDriver.prototype), 'validateDesiredCaps', this).call(this, caps);
      if (!res) {
        return res;
      }

      if (!caps.appPackage) {
        var msg = 'The desired capabilities must include an appPackage';
        _logger2['default'].errorAndThrow(msg);
      }
    }
  }, {
    key: 'proxyActive',
    value: function proxyActive(sessionId) {
      _get(Object.getPrototypeOf(TizenDriver.prototype), 'proxyActive', this).call(this, sessionId);

      return this.jwpProxyActive;
    }
  }, {
    key: 'getProxyAvoidList',
    value: function getProxyAvoidList(sessionId) {
      _get(Object.getPrototypeOf(TizenDriver.prototype), 'getProxyAvoidList', this).call(this, sessionId);

      return this.jwpProxyAvoid;
    }
  }, {
    key: 'canProxy',
    value: function canProxy(sessionId) {
      _get(Object.getPrototypeOf(TizenDriver.prototype), 'canProxy', this).call(this, sessionId);

      return false;
    }
  }, {
    key: 'appOnDevice',
    get: function get() {
      return this.helpers.isPackageOrBundle(this.opts.appPackage);
    }
  }]);

  return TizenDriver;
})(_appiumBaseDriver.BaseDriver);

exports['default'] = TizenDriver;
module.exports = exports['default'];
