// appium-ci/build/lib/ios-tools.js
'use strict';

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _utils = require('./utils');

var _utils2 = _interopRequireDefault(_utils);

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _bluebird = require('bluebird');

var _bluebird2 = _interopRequireDefault(_bluebird);

var _appiumSupport = require('appium-support');

var log = _appiumSupport.logger.getLogger('ios-tools');

var SIDE_DISK = '/Volumes/SIDE';
var SIDE_SIMS = _path2['default'].resolve(SIDE_DISK, 'sims');

function killAll() {
  var processes = arguments.length <= 0 || arguments[0] === undefined ? ['instruments', 'simulator'] : arguments[0];

  processes = _lodash2['default'].flatten([processes]);
  var seq = (0, _lodash2['default'])(processes).map(function (p) {
    return function () {
      return _utils2['default'].exec('sudo pkill -f ' + p)['catch'](function () {});
    };
  }).value();
  return _bluebird2['default'].reduce(seq, function (_, fn) {
    return fn();
  }, null);
}

function spawnAsUser(user, cmd, args) {
  log.info('running spawnAsUser', user, cmd, args);
  return _utils2['default'].exec("ps -axj | grep loginwindow | awk \"/^" + user + " / {print \\$2;exit}\"").spread(function (stdout) {
    var userPid = stdout.trim();
    return _utils2['default'].spawn("sudo", ['launchctl', 'bsexec', userPid, 'sudo', '-u'].concat([user, cmd]).concat(args), { detached: false });
  });
}

function spawnAsCurrentUser(cmd, args) {
  log.info('running spawnAsCurrentUser', cmd, args);
  return _utils2['default'].exec('whoami').spread(function (stdout) {
    var currentUser = stdout.trim();
    return spawnAsUser(currentUser, cmd, args);
  });
}

function setIosSimulatorScale() {
  log.info('setting simulator scale');
  return _utils2['default'].exec('defaults write com.apple.iphonesimulator SimulatorWindowLastScale 0.5');
}

function configureXcode(xCodeVersion) {
  log.info('configuring xCode', xCodeVersion);
  var bin = _path2['default'].resolve(SIDE_SIMS, 'configure.sh');
  return _utils2['default'].exec(bin + ' ' + xCodeVersion);
}

function resetSims() {
  log.info('resetting simulators');
  var bin = _path2['default'].resolve(SIDE_SIMS, 'reset-sims.sh');
  return _utils2['default'].exec(bin);
}

exports['default'] = {
  spawnAsUser: spawnAsUser,
  spawnAsCurrentUser: spawnAsCurrentUser,
  setIosSimulatorScale: setIosSimulatorScale,
  configureXcode: configureXcode,
  resetSims: resetSims,
  killAll: killAll
};
module.exports = exports['default'];
