// appium-base-driver/build/lib/basedriver/commands/images.js
'use strict';

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _Object$assign = require('babel-runtime/core-js/object/assign')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _protocolErrors = require('../../protocol/errors');

var _appiumSupport = require('appium-support');

var commands = {},
    helpers = {},
    extensions = {};

var MATCH_FEATURES_MODE = 'matchFeatures';
var GET_SIMILARITY_MODE = 'getSimilarity';
var MATCH_TEMPLATE_MODE = 'matchTemplate';

var DEFAULT_MATCH_THRESHOLD = 0.4;

/**
 * Performs images comparison using OpenCV framework features.
 * It is expected that both OpenCV framework and opencv4nodejs
 * module are installed on the machine where Appium server is running.
 *
 * @param {string} mode - One of possible comparison modes:
 * matchFeatures, getSimilarity, matchTemplate
 * @param {string} firstImage - Base64-encoded image file.
 * All image formats, that OpenCV library itself accepts, are supported.
 * @param {string} secondImage - Base64-encoded image file.
 * All image formats, that OpenCV library itself accepts, are supported.
 * @param {?Object} options [{}] - The content of this dictionary depends
 * on the actual `mode` value. See the documentation on `appium-support`
 * module for more details.
 * @returns {Object} The content of the resulting dictionary depends
 * on the actual `mode` and `options` values. See the documentation on
 * `appium-support` module for more details.
 * @throws {Error} If required OpenCV modules are not installed or
 * if `mode` value is incorrect or if there was an unexpected issue while
 * matching the images.
 */
commands.compareImages = function callee$0$0(mode, firstImage, secondImage) {
  var options = arguments.length <= 3 || arguments[3] === undefined ? {} : arguments[3];
  var img1, img2, result;
  return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        img1 = Buffer.from(firstImage, 'base64');
        img2 = Buffer.from(secondImage, 'base64');
        result = {};
        context$1$0.t0 = _lodash2['default'].toLower(mode);
        context$1$0.next = context$1$0.t0 === MATCH_FEATURES_MODE.toLowerCase() ? 6 : context$1$0.t0 === GET_SIMILARITY_MODE.toLowerCase() ? 10 : context$1$0.t0 === MATCH_TEMPLATE_MODE.toLowerCase() ? 14 : 18;
        break;

      case 6:
        context$1$0.next = 8;
        return _regeneratorRuntime.awrap(_appiumSupport.imageUtil.getImagesMatches(img1, img2, options));

      case 8:
        result = context$1$0.sent;
        return context$1$0.abrupt('break', 19);

      case 10:
        context$1$0.next = 12;
        return _regeneratorRuntime.awrap(_appiumSupport.imageUtil.getImagesSimilarity(img1, img2, options));

      case 12:
        result = context$1$0.sent;
        return context$1$0.abrupt('break', 19);

      case 14:
        context$1$0.next = 16;
        return _regeneratorRuntime.awrap(_appiumSupport.imageUtil.getImageOccurrence(img1, img2, options));

      case 16:
        result = context$1$0.sent;
        return context$1$0.abrupt('break', 19);

      case 18:
        throw new _protocolErrors.errors.InvalidArgumentError('\'' + mode + '\' images comparison mode is unknown. ' + ('Only ' + JSON.stringify([MATCH_FEATURES_MODE, GET_SIMILARITY_MODE, MATCH_TEMPLATE_MODE]) + ' modes are supported.'));

      case 19:
        if (!_lodash2['default'].isEmpty(result.visualization)) {
          result.visualization = result.visualization.toString('base64');
        }
        return context$1$0.abrupt('return', result);

      case 21:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
};

_Object$assign(extensions, commands, helpers);
exports.commands = commands;
exports.helpers = helpers;
exports.DEFAULT_MATCH_THRESHOLD = DEFAULT_MATCH_THRESHOLD;
exports.MATCH_TEMPLATE_MODE = MATCH_TEMPLATE_MODE;
exports['default'] = extensions;

// firstImage/img1 is the full image and secondImage/img2 is the partial one
