// appium-xcuitest-driver/build/lib/commands/recordscreen.js
'use strict';

var _createClass = require('babel-runtime/helpers/create-class')['default'];

var _classCallCheck = require('babel-runtime/helpers/class-call-check')['default'];

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _appiumSupport = require('appium-support');

var _teen_process = require('teen_process');

var _logger = require('../logger');

var _logger2 = _interopRequireDefault(_logger);

var _utils = require('../utils');

var _wdaIproxy = require('../wda/iproxy');

var _wdaIproxy2 = _interopRequireDefault(_wdaIproxy);

var commands = {};

var MAX_RECORDING_TIME_SEC = 60 * 30;
var DEFAULT_RECORDING_TIME_SEC = 60 * 3;
var DEFAULT_MJPEG_SERVER_PORT = 9100;
var MP4_EXT = '.mp4';
var DEFAULT_FPS = 10;
var FFMPEG_BINARY = 'ffmpeg';
var ffmpegLogger = _appiumSupport.logger.getLogger(FFMPEG_BINARY);

var ScreenRecorder = (function () {
  function ScreenRecorder(udid, videoPath) {
    var opts = arguments.length <= 2 || arguments[2] === undefined ? {} : arguments[2];

    _classCallCheck(this, ScreenRecorder);

    this.videoPath = videoPath;
    this.opts = opts;
    this.udid = udid;
    this.mainProcess = null;
    this.iproxy = null;
    this.timeoutHandler = null;
  }

  /**
   * @typedef {Object} StartRecordingOptions
   *
   * @property {?string} remotePath - The path to the remote location, where the resulting video should be uploaded.
   *                                  The following protocols are supported: http/https, ftp.
   *                                  Null or empty string value (the default setting) means the content of resulting
   *                                  file should be encoded as Base64 and passed as the endpount response value.
   *                                  An exception will be thrown if the generated media file is too big to
   *                                  fit into the available process memory.
   *                                  This option only has an effect if there is screen recording process in progreess
   *                                  and `forceRestart` parameter is not set to `true`.
   * @property {?string} user - The name of the user for the remote authentication. Only works if `remotePath` is provided.
   * @property {?string} pass - The password for the remote authentication. Only works if `remotePath` is provided.
   * @property {?string} method - The http multipart upload method name. The 'PUT' one is used by default.
   *                              Only works if `remotePath` is provided.
   * @property {?string} videoType - The format of the screen capture to be recorded.
   *                                 Available formats: "h264", "mp4" or "fmp4". Default is "mp4".
   * @property {?string} videoQuality - The video encoding quality (low, medium, high, photo - defaults to medium).
   * @property {?string} videoFps - The Frames Per Second rate of the recorded video. Change this value if the resulting video
   *                                is too slow or too fast. Defaults to 10.
   * @property {?boolean} forceRestart - Whether to try to catch and upload/return the currently running screen recording
   *                                     (`false`, the default setting) or ignore the result of it and start a new recording
   *                                     immediately.
   * @property {?string|number} timeLimit - The maximum recording time, in seconds.
   *                                        The default value is 180, the maximum value is 600 (10 minutes).
   */

  /**
   * Record the display of devices running iOS Simulator since Xcode 8.3 or real devices since iOS 8
   * (ios-minicap utility is required: https://github.com/openstf/ios-minicap).
   * It records screen activity to a MPEG-4 file. Audio is not recorded with the video file.
   * If screen recording has been already started then the command will stop it forcefully and start a new one.
   * The previously recorded video file will be deleted.
   *
   * @param {?StartRecordingOptions} options - The available options.
   * @returns {string} Base64-encoded content of the recorded media file if
   *                   any screen recording is currently running or an empty string.
   * @throws {Error} If screen recording has failed to start.
   */

  _createClass(ScreenRecorder, [{
    key: 'start',
    value: function start(timeout) {
      var localPort, args;
      return _regeneratorRuntime.async(function start$(context$2$0) {
        var _this = this;

        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            context$2$0.prev = 0;
            context$2$0.next = 3;
            return _regeneratorRuntime.awrap(_appiumSupport.fs.which(FFMPEG_BINARY));

          case 3:
            context$2$0.next = 8;
            break;

          case 5:
            context$2$0.prev = 5;
            context$2$0.t0 = context$2$0['catch'](0);
            throw new Error('\'' + FFMPEG_BINARY + '\' binary is not found in PATH. Install it using \'brew install ffmpeg\'. ' + 'Check https://www.ffmpeg.org/download.html for more details.');

          case 8:
            localPort = this.opts.remotePort;

            if (!this.opts.usePortForwarding) {
              context$2$0.next = 12;
              break;
            }

            context$2$0.next = 12;
            return _regeneratorRuntime.awrap(this.startIproxy(localPort));

          case 12:
            args = ['-f', 'mjpeg', '-r', this.opts.videoFps, '-i', 'http://localhost:' + localPort, '-vcodec', 'mjpeg', '-y', this.videoPath];

            this.mainProcess = new _teen_process.SubProcess(FFMPEG_BINARY, args);
            this.mainProcess.on('output', function (stdout, stderr) {
              if (stderr && !stderr.includes('frame=')) {
                ffmpegLogger.info('' + stderr);
              }
            });
            // Give ffmpeg some time for init
            context$2$0.next = 17;
            return _regeneratorRuntime.awrap(this.mainProcess.start(5000));

          case 17:
            _logger2['default'].info('Starting screen capture on the device \'' + this.udid + '\' with command: \'' + FFMPEG_BINARY + ' ' + args.join(' ') + '\'. ' + ('Will timeout in ' + timeout + 'ms'));

            this.timeoutHandler = setTimeout(function callee$2$0() {
              return _regeneratorRuntime.async(function callee$2$0$(context$3$0) {
                while (1) switch (context$3$0.prev = context$3$0.next) {
                  case 0:
                    context$3$0.next = 2;
                    return _regeneratorRuntime.awrap(this.interrupt());

                  case 2:
                    if (context$3$0.sent) {
                      context$3$0.next = 4;
                      break;
                    }

                    _logger2['default'].warn('Cannot finish the active screen recording on the device \'' + this.udid + '\' after ' + timeout + 'ms timeout');

                  case 4:
                  case 'end':
                    return context$3$0.stop();
                }
              }, null, _this);
            }, timeout);

          case 19:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this, [[0, 5]]);
    }
  }, {
    key: 'startIproxy',
    value: function startIproxy(localPort) {
      return _regeneratorRuntime.async(function startIproxy$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            this.iproxy = new _wdaIproxy2['default'](this.udid, localPort, this.opts.remotePort);
            context$2$0.prev = 1;
            context$2$0.next = 4;
            return _regeneratorRuntime.awrap(this.iproxy.start());

          case 4:
            context$2$0.next = 10;
            break;

          case 6:
            context$2$0.prev = 6;
            context$2$0.t0 = context$2$0['catch'](1);

            _logger2['default'].warn('Cannot start iproxy. Assuming it is already forwarding the remote port ' + this.opts.remotePort + ' to ' + localPort + ' ' + ('for the device ' + this.udid + '. Set the custom value to \'mjpegServerPort\' capability if this is an undesired behavior. ') + ('Original error: ' + context$2$0.t0.message));
            this.iproxy = null;

          case 10:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this, [[1, 6]]);
    }
  }, {
    key: 'stopIproxy',
    value: function stopIproxy() {
      var quitPromise;
      return _regeneratorRuntime.async(function stopIproxy$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            if (this.iproxy) {
              context$2$0.next = 2;
              break;
            }

            return context$2$0.abrupt('return');

          case 2:
            quitPromise = this.iproxy.quit();

            this.iproxy = null;
            context$2$0.prev = 4;
            context$2$0.next = 7;
            return _regeneratorRuntime.awrap(quitPromise);

          case 7:
            context$2$0.next = 12;
            break;

          case 9:
            context$2$0.prev = 9;
            context$2$0.t0 = context$2$0['catch'](4);

            _logger2['default'].warn('Cannot stop iproxy. Original error: ' + context$2$0.t0.message);

          case 12:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this, [[4, 9]]);
    }
  }, {
    key: 'interrupt',
    value: function interrupt() {
      var force = arguments.length <= 0 || arguments[0] === undefined ? false : arguments[0];
      var result, interruptPromise;
      return _regeneratorRuntime.async(function interrupt$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            result = true;

            if (this.timeoutHandler) {
              clearTimeout(this.timeoutHandler);
              this.timeoutHandler = null;
            }

            if (!(this.mainProcess && this.mainProcess.isRunning)) {
              context$2$0.next = 14;
              break;
            }

            interruptPromise = this.mainProcess.stop(force ? 'SIGTERM' : 'SIGINT');

            this.mainProcess = null;
            context$2$0.prev = 5;
            context$2$0.next = 8;
            return _regeneratorRuntime.awrap(interruptPromise);

          case 8:
            context$2$0.next = 14;
            break;

          case 10:
            context$2$0.prev = 10;
            context$2$0.t0 = context$2$0['catch'](5);

            _logger2['default'].warn('Cannot ' + (force ? 'terminate' : 'interrupt') + ' ' + FFMPEG_BINARY + '. ' + ('Original error: ' + context$2$0.t0.message));
            result = false;

          case 14:
            if (!this.opts.usePortForwarding) {
              context$2$0.next = 17;
              break;
            }

            context$2$0.next = 17;
            return _regeneratorRuntime.awrap(this.stopIproxy());

          case 17:
            return context$2$0.abrupt('return', result);

          case 18:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this, [[5, 10]]);
    }
  }, {
    key: 'finish',
    value: function finish() {
      return _regeneratorRuntime.async(function finish$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            context$2$0.next = 2;
            return _regeneratorRuntime.awrap(this.interrupt());

          case 2:
            return context$2$0.abrupt('return', this.videoPath);

          case 3:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'cleanup',
    value: function cleanup() {
      return _regeneratorRuntime.async(function cleanup$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            context$2$0.next = 2;
            return _regeneratorRuntime.awrap(_appiumSupport.fs.exists(this.videoPath));

          case 2:
            if (!context$2$0.sent) {
              context$2$0.next = 5;
              break;
            }

            context$2$0.next = 5;
            return _regeneratorRuntime.awrap(_appiumSupport.fs.rimraf(this.videoPath));

          case 5:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }]);

  return ScreenRecorder;
})();

commands.startRecordingScreen = function callee$0$0() {
  var options = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];

  var videoType, _options$timeLimit, timeLimit, _options$videoQuality, videoQuality, _options$videoFps, videoFps, forceRestart, result, videoPath, screenRecorder, timeout;

  return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        videoType = options.videoType;
        _options$timeLimit = options.timeLimit;
        timeLimit = _options$timeLimit === undefined ? DEFAULT_RECORDING_TIME_SEC : _options$timeLimit;
        _options$videoQuality = options.videoQuality;
        videoQuality = _options$videoQuality === undefined ? 'medium' : _options$videoQuality;
        _options$videoFps = options.videoFps;
        videoFps = _options$videoFps === undefined ? DEFAULT_FPS : _options$videoFps;
        forceRestart = options.forceRestart;
        result = '';

        if (forceRestart) {
          context$1$0.next = 14;
          break;
        }

        _logger2['default'].info('Checking if there is/was a previous screen recording. ' + 'Set \'forceRestart\' option to \'true\' if you\'d like to skip this step.');
        context$1$0.next = 13;
        return _regeneratorRuntime.awrap(this.stopRecordingScreen(options));

      case 13:
        result = context$1$0.sent;

      case 14:
        context$1$0.next = 16;
        return _regeneratorRuntime.awrap(_appiumSupport.tempDir.path({
          prefix: 'appium_' + Math.random().toString(16).substring(2, 8),
          suffix: MP4_EXT
        }));

      case 16:
        videoPath = context$1$0.sent;
        screenRecorder = new ScreenRecorder(this.opts.device.udid, videoPath, {
          // TODO: Apply type and quality options
          videoType: videoType,
          videoQuality: videoQuality,
          videoFps: videoFps || DEFAULT_FPS,
          remotePort: this.opts.mjpegServerPort || DEFAULT_MJPEG_SERVER_PORT,
          usePortForwarding: this.isRealDevice()
        });
        context$1$0.next = 20;
        return _regeneratorRuntime.awrap(screenRecorder.interrupt(true));

      case 20:
        if (context$1$0.sent) {
          context$1$0.next = 22;
          break;
        }

        _logger2['default'].errorAndThrow('Unable to stop screen recording process');

      case 22:
        if (!this._recentScreenRecorder) {
          context$1$0.next = 26;
          break;
        }

        context$1$0.next = 25;
        return _regeneratorRuntime.awrap(this._recentScreenRecorder.cleanup());

      case 25:
        this._recentScreenRecorder = null;

      case 26:
        timeout = parseFloat(timeLimit);

        if (isNaN(timeout) || timeout > MAX_RECORDING_TIME_SEC || timeout <= 0) {
          _logger2['default'].errorAndThrow('The timeLimit value must be in range [1, ' + MAX_RECORDING_TIME_SEC + '] seconds. ' + ('The value of \'' + timeLimit + '\' has been passed instead.'));
        }

        context$1$0.prev = 28;
        context$1$0.next = 31;
        return _regeneratorRuntime.awrap(screenRecorder.start(timeout));

      case 31:
        context$1$0.next = 40;
        break;

      case 33:
        context$1$0.prev = 33;
        context$1$0.t0 = context$1$0['catch'](28);
        context$1$0.next = 37;
        return _regeneratorRuntime.awrap(screenRecorder.interrupt(true));

      case 37:
        context$1$0.next = 39;
        return _regeneratorRuntime.awrap(screenRecorder.cleanup());

      case 39:
        throw context$1$0.t0;

      case 40:
        this._recentScreenRecorder = screenRecorder;

        return context$1$0.abrupt('return', result);

      case 42:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this, [[28, 33]]);
};

/**
 * @typedef {Object} StopRecordingOptions
 *
 * @property {?string} remotePath - The path to the remote location, where the resulting video should be uploaded.
 *                                  The following protocols are supported: http/https, ftp.
 *                                  Null or empty string value (the default setting) means the content of resulting
 *                                  file should be encoded as Base64 and passed as the endpount response value.
 *                                  An exception will be thrown if the generated media file is too big to
 *                                  fit into the available process memory.
 * @property {?string} user - The name of the user for the remote authentication. Only works if `remotePath` is provided.
 * @property {?string} pass - The password for the remote authentication. Only works if `remotePath` is provided.
 * @property {?string} method - The http multipart upload method name. The 'PUT' one is used by default.
 *                              Only works if `remotePath` is provided.
 */

/**
 * Stop recording the screen. If no screen recording process is running then
 * the endpoint will try to get the recently recorded file.
 * If no previously recorded file is found and no active screen recording
 * processes are running then the method returns an empty string.
 *
 * @param {?StopRecordingOptions} options - The available options.
 * @returns {string} Base64-encoded content of the recorded media file if 'remotePath'
 *                   parameter is empty or null or an empty string.
 * @throws {Error} If there was an error while getting the name of a media file
 *                 or the file content cannot be uploaded to the remote location.
 */
commands.stopRecordingScreen = function callee$0$0() {
  var options = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];
  var remotePath, user, pass, method, videoPath;
  return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        remotePath = options.remotePath;
        user = options.user;
        pass = options.pass;
        method = options.method;

        if (this._recentScreenRecorder) {
          context$1$0.next = 7;
          break;
        }

        _logger2['default'].info('Screen recording is not running. There is nothing to stop.');
        return context$1$0.abrupt('return', '');

      case 7:
        context$1$0.prev = 7;
        context$1$0.next = 10;
        return _regeneratorRuntime.awrap(this._recentScreenRecorder.finish());

      case 10:
        videoPath = context$1$0.sent;
        context$1$0.next = 13;
        return _regeneratorRuntime.awrap(_appiumSupport.fs.exists(videoPath));

      case 13:
        if (context$1$0.sent) {
          context$1$0.next = 15;
          break;
        }

        _logger2['default'].errorAndThrow('The screen recorder utility has failed ' + ('to store the actual screen recording at \'' + videoPath + '\''));

      case 15:
        context$1$0.next = 17;
        return _regeneratorRuntime.awrap((0, _utils.encodeBase64OrUpload)(videoPath, remotePath, {
          user: user,
          pass: pass,
          method: method
        }));

      case 17:
        return context$1$0.abrupt('return', context$1$0.sent);

      case 18:
        context$1$0.prev = 18;
        context$1$0.next = 21;
        return _regeneratorRuntime.awrap(this._recentScreenRecorder.interrupt(true));

      case 21:
        context$1$0.next = 23;
        return _regeneratorRuntime.awrap(this._recentScreenRecorder.cleanup());

      case 23:
        this._recentScreenRecorder = null;
        return context$1$0.finish(18);

      case 25:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this, [[7,, 18, 25]]);
};

exports.commands = commands;
exports['default'] = commands;
